realtek rtl8195am SLEEP feature implementation and lib updates

M-ichae-l opened the pull request:
### Description

1, Rtl8195am lib updates. Sync with the latest rtl8195am SDK.
2, Add sleep feature. Add "sleep_ex_api.h", "sleep.c" and "sleep_api.c".
3, Edit api files to fit the new lib updates.

### Pull request type
    [X] Feature

M-ichae-l committed:
realtek rtl8195am SLEEP features implementation and lib updates

1, lib updates. Sync with the latest rtl8195am SDK
2, Add sleep features. Add "sleep_ex_api.h", "sleep.c" and "sleep_api.c"
3, edit api files to fit new lib

0xc0170 commented on targets/TARGET_Realtek/TARGET_AMEBA/TARGET_RTL8195A/lp_ticker_api.c:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    // Initial a G-Timer
+#if 0
```
Dead code, please remove

0xc0170 commented on targets/TARGET_Realtek/TARGET_AMEBA/TARGET_RTL8195A/lp_ticker_api.c:
```diff
@@ -0,0 +1,269 @@
+ * }
+ * @endcode
+ */
+#if 1
```
This `#if ` should also be cleaned up

0xc0170 reviewed (changes requested):
Can you also add license for ram_1.bin ?

M-ichae-l committed:
realtek rtl8195am SLEEP features implementation and lib updates

1, remove the extra"lp_ticker_api.c"

M-ichae-l commented:
@0xc0170  
"lp_ticker_api.c" is an extra file. So remove it for this PR.

M-ichae-l commented:
@0xc0170 
For the ram_1.bin, could you give some suggestions about how to add the license?

0xc0170 commented:
Add `LICENSE` file and specify binary license. I assume this would be `Permissive Binary License` ?

M-ichae-l commented:
@0xc0170 
Thank you!
I have checked the `bootloaders file`. There are 4 targets and all of them do not have `LICENSE` (including RTL8195AM, previously from the master branch). Are you going to request all targets to add LICENSE or just the RTL8195AM?

0xc0170 commented:
I was going to create a tracking issue for this, will do asap (they should have).

Update: Tracking issue created for the rest of the bootloaders there.

M-ichae-l committed:
realtek rtl8195am SLEEP features implementation and lib updates

1, Add license for bootloader

M-ichae-l commented:
@0xc0170 
I have committed a license for bootloader. Could you double check if the license is acceptable?

0xc0170 commented:
Thanks, looks good to me

0xc0170 reviewed (approved):
There are few libraries updated here, and I checked if they have license (they don't). Can you please send a separate pull request adding the license for those libraries (driver/wlan) ?

0xc0170 commented:
/morph build

0xc0170 commented:
To be specific: lib_peripheral and lib_wlan libs - license

cmonr commented:
@M-ichae-l Would you mind taking a look at the tests-mbed_hal-sleep test failures?

M-ichae-l commented:
@cmonr 
The tests-mbed_hal-sleep test failures are related to `sleep_usticker_test`. However, the new us_ticker (also lp_ticker) features have not been implemented yet. The failures are expected to appear.

cmonr commented:
Interesting, and thanks for the additional info.

Do you have any idea if fixes will be coming into this PR for them, or if there's another PR/issue that is being worked on to resolve them? I'm afraid we can't bring something in that we know would fail tests.

M-ichae-l commented:
@cmonr 
Let me apply fixes in this PR to pass the `sleep_usticker_test`. Thank you! 
The new us ticker and lp ticker features will be updated in another coming PR once it is ready.

M-ichae-l committed:
realtek rtl8195am SLEEP features implementation and lib updates

1, replace "us_ticker.c" with "us_ticker_api.c". To correct the return of "us_ticker_read(void)" and pass the "sleep_usticker_test"

M-ichae-l commented:
@cmonr 
I have use a new `us_ticker_api.c` to replace the old `us_ticker.c`. The new `us_ticker_api.c` fix return of `us_ticker_read(void)` to be ticks. This fix will let the `sleep_usticker_test` pass.

0xc0170 commented:
There's conflict now, please resolve

M-ichae-l commented:
@0xc0170 
I don't have the write access to resolve the conflicts. Could I have the access? Also, I did not see the conflicts offline. Please let me see the conflicts online.

0xc0170 commented:
`M-ichae-l:realtek-rtl8195am-SLEEP-features-implementation` - this is where to resolve the conflicts. You shall rebase this branch on top of master

I would do:

```
git checkout master
git pull --rebase upstream master #upstream is Mbed OS git repository
git checkout realtek-rtl8195am-SLEEP-features-implementation
git rebase master # right now will display conflicts and asks you to resolve them
```

M-ichae-l commented:
@0xc0170 
Please refer to #7715 for the rebase and resolved conflicts. 

We can close this PR.
